File: scripts/smoke_test.py
"""
Opens the Zenith TUI in a visible window + sends 'Hi' via WebSocket and logs the response.
You see the app visually AND get a clean log of everything that happened.

Usage:
    python scripts/smoke_test.py
    python scripts/smoke_test.py --prompt "explain compaction"
    python scripts/smoke_test.py --wait 8
"""

import argparse
import asyncio
import json
import os
import subprocess
import sys
import time
import uuid
from datetime import datetime
from pathlib import Path

try:
    import websockets
except ImportError:
    subprocess.check_call([sys.executable, "-m", "pip", "install", "websockets", "-q"])
    import websockets


WS_URL = "ws://127.0.0.1:8765/ws"


def ts():
    now = datetime.now()
    return now.strftime("%H:%M:%S.") + f"{now.microsecond // 1000:03d}"


def format_event(msg: dict) -> str:
    """Format a single event into a readable line."""
    params = msg.get("params", {})
    kind = params.get("kind", "?")
    data = params.get("data", {})

    if kind == "message":
        return data.get("text", "")
    elif kind == "tool_call":
        tool = data.get("tool", "?")
        p = json.dumps(data.get("params", {}))
        if len(p) > 120:
            p = p[:120] + "..."
        return f"  [{tool}] {p}"
    elif kind == "tool_result":
        tool = data.get("tool", "?")
        ok = data.get("success", "?")
        err = data.get("error", "")
        out = data.get("output", "") or ""
        if err:
            return f"  [{tool}] ERROR: {err[:200]}"
        return f"  [{tool}] OK ({len(out)} chars)"
    elif kind == "error":
        return f"  ERROR: {data.get('message', data)}"
    elif kind == "warning":
        return f"  WARNING: {data.get('message', data)}"
    elif kind == "thinking":
        text = data.get("text", "")
        dur = data.get("duration", 0)
        return f"  [thinking {dur}ms] {text[:200]}"
    elif kind == "progress":
        return f"  ... {data.get('label', '')}"
    elif kind == "turn_manifest":
        return f"  verdict={data.get('verdict', '?')}"
    elif kind == "context_compacted":
        return f"  [compacted] {json.dumps(data)[:150]}"
    else:
        return f"  [{kind}] {json.dumps(data)[:200]}"


async def run(prompt: str, wait: int, response_timeout: int):
    workspace = Path(__file__).resolve().parent.parent
    log_name = f"smoke_test_{time.strftime('%Y%m%d_%H%M%S')}.log"
    log_path = workspace / log_name
    out: list[str] = []

    def log(msg=""):
        line = f"[{ts()}] {msg}" if msg else ""
        out.append(line)
        print(line)

    log("=" * 70)
    log("SMOKE TEST")
    log(f"Prompt : {prompt!r}")
    log(f"Log    : {log_path}")
    log("=" * 70)

    # ── Kill old server ───────────────────────────────────────────────
    log("\n[1] Cleaning up...")
    my_pid = os.getpid()
    try:
        r = subprocess.run(["netstat", "-ano"], capture_output=True, text=True,
                           creationflags=0x08000000)
        for line in r.stdout.splitlines():
            if ":8765" in line and "LISTENING" in line:
                pid = int(line.split()[-1])
                if pid != my_pid:
                    subprocess.run(["taskkill", "/F", "/PID", str(pid)],
                                   capture_output=True, creationflags=0x08000000)
                    log(f"  Killed old server PID {pid}")
    except Exception:
        pass

    # ── Start backend ─────────────────────────────────────────────────
    log("[2] Starting backend...")
    venv_py = workspace / ".venv" / "Scripts" / "python.exe"
    if not venv_py.exists():
        venv_py = Path(sys.executable)
    server = subprocess.Popen(
        [str(venv_py), "-m", "server.main", "serve"],
        cwd=str(workspace),
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
        creationflags=0x08000000,
    )
    log(f"  PID={server.pid}, waiting 4s...")
    await asyncio.sleep(4)

    # ── Open TUI visually ─────────────────────────────────────────────
    log("[3] Opening TUI in new terminal window (for you to see)...")
    tui_cmd = f'cd /d "{workspace}" && title Zenith-SmokeTest && npm run dev:frontend'
    subprocess.Popen(
        ["cmd", "/c", "start", "cmd", "/k", tui_cmd],
        cwd=str(workspace),
    )
    log(f"  Waiting {wait}s for TUI to initialize...")
    await asyncio.sleep(wait)

    # ── Connect via WebSocket ─────────────────────────────────────────
    log("[4] Connecting to backend via WebSocket...")
    async with websockets.connect(WS_URL, max_size=10 * 1024 * 1024) as ws:
        log("  Connected.\n")
        log(f"  > {prompt}")
        log("  " + "-" * 50)

        # Create session
        sid = str(uuid.uuid4())[:8]
        await ws.send(json.dumps({
            "jsonrpc": "2.0", "id": "s1",
            "method": "session.create",
            "params": {"title": f"smoke-{sid}"},
        }))
        session_id = None
        for _ in range(20):
            try:
                raw = await asyncio.wait_for(ws.recv(), timeout=1)
                msg = json.loads(raw)
                if msg.get("id") == "s1" and "result" in msg:
                    session_id = msg["result"].get("session_id") or msg["result"].get("id")
                    break
            except asyncio.TimeoutError:
                continue

        if not session_id:
            log("  ERROR: session.create failed")
            _save(log_path, out)
            return

        # Send prompt
        await ws.send(json.dumps({
            "jsonrpc": "2.0", "id": "p1",
            "method": "prompt.send",
            "params": {"content": prompt, "mode": "build", "session_id": session_id},
        }))

        # Collect events
        log("")
        full_answer = []
        start = time.time()
        done = False

        while time.time() - start < response_timeout:
            try:
                raw = await asyncio.wait_for(ws.recv(), timeout=2)
                msg = json.loads(raw)

                if msg.get("method") == "event":
                    line = format_event(msg)
                    if line:
                        log(line)
                    kind = msg.get("params", {}).get("kind", "")
                    text = msg.get("params", {}).get("data", {}).get("text", "")
                    if kind == "message" and text:
                        full_answer.append(text)
                    if kind == "turn_manifest":
                        v = msg["params"]["data"].get("verdict", "")
                        if v in ("completed", "answered"):
                            done = True
                            break

            except asyncio.TimeoutError:
                if done:
                    break
                continue

        elapsed = time.time() - start
        log("  " + "-" * 50)
        log(f"\n  ANSWER ({elapsed:.1f}s):")
        log("  " + "".join(full_answer))
        log("")

    # ── Save ──────────────────────────────────────────────────────────
    _save(log_path, out)
    log(f"Log saved: {log_path}")
    print(f"\n{'='*70}")
    print(f"  LOG: {log_path}")
    print(f"{'='*70}")


def _save(path, lines):
    with open(path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines) + "\n")


def main():
    p = argparse.ArgumentParser(description="Open TUI + send prompt via WebSocket")
    p.add_argument("--prompt", default="Hi")
    p.add_argument("--wait", type=int, default=10, help="Secs to wait for TUI to load")
    p.add_argument("--response-timeout", type=int, default=120)
    args = p.parse_args()
    asyncio.run(run(args.prompt, args.wait, args.response_timeout))


if __name__ == "__main__":
    main()

File: scripts/smoke_test.js
/**
 * Smoke test: uses node-pty to spawn the TUI, type "Hi", capture all output.
 *
 * Usage:
 *   node scripts/smoke_test.js
 *   node scripts/smoke_test.js --prompt "explain compaction"
 *   node scripts/smoke_test.js --wait 8000
 */

const pty = require("node-pty");
const { spawn, execSync } = require("child_process");
const fs = require("fs");
const path = require("path");

const WORKSPACE = path.resolve(__dirname, "..");
const LOG_DIR = WORKSPACE;

// Parse args
const args = process.argv.slice(2);
function getArg(name, fallback) {
  const idx = args.indexOf(name);
  return idx !== -1 && args[idx + 1] ? args[idx + 1] : fallback;
}
const PROMPT = getArg("--prompt", "Hi");
const WAIT_BEFORE_INPUT = parseInt(getArg("--wait", "12000"), 10);
const RESPONSE_TIMEOUT = parseInt(getArg("--timeout", "60000"), 10);

const logFile = path.join(LOG_DIR, `smoke_test_${timestamp()}.log`);
const logLines = [];

function timestamp() {
  const d = new Date();
  return (
    d.getFullYear().toString() +
    String(d.getMonth() + 1).padStart(2, "0") +
    String(d.getDate()).padStart(2, "0") + "_" +
    String(d.getHours()).padStart(2, "0") +
    String(d.getMinutes()).padStart(2, "0") +
    String(d.getSeconds()).padStart(2, "0")
  );
}

function logTime() {
  const d = new Date();
  return (
    String(d.getHours()).padStart(2, "0") + ":" +
    String(d.getMinutes()).padStart(2, "0") + ":" +
    String(d.getSeconds()).padStart(2, "0") + "." +
    String(d.getMilliseconds()).padStart(3, "0")
  );
}

function log(msg = "") {
  const line = `[${logTime()}] ${msg}`;
  logLines.push(line);
  process.stdout.write(line + "\n");
}

function saveLog() {
  fs.writeFileSync(logFile, logLines.join("\n") + "\n", "utf-8");
}

// Strip ANSI escape codes for clean log
function stripAnsi(str) {
  return str
    // eslint-disable-next-line no-control-regex
    .replace(
      /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><~]/g,
      ""
    )
    .replace(/\r/g, "");
}

async function main() {
  log("=" .repeat(70));
  log("SMOKE TEST: node-pty TUI automation");
  log(`Prompt : ${PROMPT}`);
  log(`Log    : ${logFile}`);
  log("=" .repeat(70));

  // ── 1. Kill old server ─────────────────────────────────────────────
  log("\n[1/3] Cleaning up old server...");
  try {
    const netstat = execSync("netstat -ano", { encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] });
    const myPid = process.pid;
    for (const line of netstat.split("\n")) {
      if (line.includes(":8765") && line.includes("LISTENING")) {
        const pid = parseInt(line.trim().split(/\s+/).pop(), 10);
        if (pid && pid !== myPid) {
          try {
            execSync(`taskkill /F /PID ${pid}`, { stdio: "pipe" });
            log(`  Killed old server PID ${pid}`);
          } catch {}
        }
      }
    }
  } catch {}

  // ── 2. Start backend + health check ────────────────────────────────
  log("[2/3] Starting backend server...");
  const venvPy = path.join(WORKSPACE, ".venv", "Scripts", "python.exe");
  const python = fs.existsSync(venvPy) ? venvPy : "python";
  const serverProc = spawn(python, ["-m", "server.main", "serve"], {
    cwd: WORKSPACE,
    stdio: "ignore",
    detached: true,
    windowsHide: true,
  });
  serverProc.unref();
  log(`  Server PID: ${serverProc.pid}`);

  // Health check: poll until server responds or timeout
  const http = require("http");
  const HEALTH_TIMEOUT = 15000;
  const HEALTH_INTERVAL = 500;
  const healthStart = Date.now();
  let ready = false;

  log("  Waiting for server to be ready...");
  while (Date.now() - healthStart < HEALTH_TIMEOUT) {
    try {
      await new Promise((resolve, reject) => {
        const req = http.get("http://127.0.0.1:8765/health", (res) => {
          resolve(res.statusCode);
        });
        req.on("error", reject);
        req.setTimeout(1000, () => { req.destroy(); reject(new Error("timeout")); });
      });
      ready = true;
      break;
    } catch {
      await sleep(HEALTH_INTERVAL);
    }
  }

  if (!ready) {
    log(`  WARNING: Server not ready after ${HEALTH_TIMEOUT}ms, proceeding anyway.`);
  } else {
    log(`  Server ready in ${Date.now() - healthStart}ms.`);
  }

  // ── 3. Spawn TUI via node-pty ──────────────────────────────────────
  log("[3/3] Spawning TUI via node-pty...");

  const shell = process.env.COMSPEC || "cmd.exe";
  // Use npx which is always on PATH
  const batchCmd = `cd /d ${WORKSPACE} && npx tsx tui/src/index.tsx`;
  log(`  Command: ${batchCmd}`);

  const ptyProc = pty.spawn(shell, ["/c", batchCmd], {
    name: "xterm-256color",
    cols: 160,
    rows: 50,
    cwd: WORKSPACE,
    env: { ...process.env, TERM: "xterm-256color" },
  });

  log(`  PTY PID: ${ptyProc.pid}`);
  log(`  Waiting ${WAIT_BEFORE_INPUT}ms for TUI to load...`);

  // Collect all output
  let allOutput = "";
  let cleanOutput = "";
  ptyProc.onData((data) => {
    allOutput += data;
    cleanOutput += stripAnsi(data);
  });

  await sleep(WAIT_BEFORE_INPUT);
  log("  TUI should be loaded.");

  // ── 4. Type prompt ─────────────────────────────────────────────────
  log(`\n--- TYPING: ${PROMPT} ---`);
  ptyProc.write(PROMPT);
  await sleep(300);
  ptyProc.write("\r");  // Enter
  log("  Sent.\n");

  // ── 5. Wait for response ───────────────────────────────────────────
  log(`--- WAITING ${RESPONSE_TIMEOUT}ms FOR RESPONSE ---`);

  const startTime = Date.now();
  let lastLen = 0;
  let stableCount = 0;

  while (Date.now() - startTime < RESPONSE_TIMEOUT) {
    await sleep(1000);
    const currentLen = cleanOutput.length;

    if (currentLen === lastLen) {
      stableCount++;
    } else {
      stableCount = 0;
      lastLen = currentLen;
    }

    // If output stabilized for 8 seconds after we got some content, we're done
    if (stableCount >= 8 && currentLen > 100) {
      log("  Output stabilized — response complete.");
      break;
    }

    const elapsed = Math.round((Date.now() - startTime) / 1000);
    process.stdout.write(`\r  ${elapsed}s / ${Math.round(RESPONSE_TIMEOUT / 1000)}s — ${currentLen} chars...`);
  }
  process.stdout.write("\r" + " ".repeat(60) + "\r");

  // ── 6. Log captured output ─────────────────────────────────────────
  log("\n" + "=".repeat(70));
  log("CAPTURED TUI OUTPUT (clean, no ANSI):");
  log("=".repeat(70));
  log("");

  const lines = cleanOutput.split("\n");
  for (const line of lines) {
    const trimmed = line.trimEnd();
    if (trimmed.length > 0) {
      log(`  ${trimmed}`);
    }
  }

  log("");
  log("=".repeat(70));
  log(`Total: ${cleanOutput.length} chars, ${lines.length} lines`);
  log("=".repeat(70));

  // ── 7. Cleanup ─────────────────────────────────────────────────────
  try { ptyProc.kill(); } catch {}
  try { execSync("taskkill /F /PID " + serverProc.pid, { stdio: "pipe" }); } catch {}
  try { execSync("taskkill /F /IM node.exe /FI \"WINDOWTITLE eq *smoke*\"", { stdio: "pipe" }); } catch {}

  saveLog();

  console.log(`\n${"=".repeat(70)}`);
  console.log(`  LOG FILE: ${logFile}`);
  console.log(`${"=".repeat(70)}\n`);
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  console.error("FATAL:", err);
  saveLog();
  process.exit(1);
});
